Firestore/core/src/firebase/firestore/remote/grpc_stream.h

```diff
@@ -0,0 +1,238 @@
+ *
+ * The stream is disposable; once it finishes, it cannot be restarted.
+ *
+ * This class is essentially a wrapper over
+ * `grpc::GenericClientAsyncReaderWriter`.
+ */
+class GrpcStream : public std::enable_shared_from_this<GrpcStream> {
+ public:
+  // Implementation of the stream relies on its memory being managed by
+  // `shared_ptr`.
+  //
+  // The given `grpc_queue` must wrap the same underlying completion queue as
+  // the `call`.
+  static std::shared_ptr<GrpcStream> MakeStream(
+      std::unique_ptr<grpc::ClientContext> context,
+      std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+      GrpcStreamObserver* observer,
+      GrpcCompletionQueue* grpc_queue);
+
+  void Start();
+
+  void Write(grpc::ByteBuffer&& message);
+
+  // Does not produce a notification. Once this method is called, the stream can
+  // no longer be used.
+  //
+  // Can be called on a stream before it opens. It is invalid to finish a stream
+  // more than once.
+  void Finish();
+
+  /**
+   * Writes the given message and finishes the stream as soon as the write
+   * succeeds. Any non-started writes will be discarded. Neither write nor
+   * finish will notify the observer.
+   *
+   * If the stream hasn't opened yet, `WriteAndFinish` is equivalent to
+   * `Finish` -- the write will be ignored.
+   */
+  void WriteAndFinish(grpc::ByteBuffer&& message);
+
+ private:
+  friend class internal::GrpcStreamDelegate;
+
+  GrpcStream(std::unique_ptr<grpc::ClientContext> context,
+             std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+             GrpcStreamObserver* observer,
+             GrpcCompletionQueue* grpc_queue);
+
+  void Read();
+  void BufferedWrite(grpc::ByteBuffer&& message);
+
+  // Called by `GrpcStreamDelegate`.
+  void OnStart();
+  void OnRead(const grpc::ByteBuffer& message);
+  void OnWrite();
+  void OnOperationFailed();
+  void OnFinishedByServer(const grpc::Status& status);
+  void OnFinishedByClient();
+  void RemoveOperation(const internal::StreamOperation* to_remove);
+
+  // Whether this stream belongs to the same generation as the observer.
+  bool SameGeneration() const;
+
+  // Creates an operation, stores a `unique_ptr` to it, and returns a plain
+  // pointer.
+  template <typename Op, typename... Args>
+  Op* MakeOperation(Args... args);
+
+  // Creates and immediately executes an operation.
+  template <typename Op, typename... Args>
+  void Execute(Args... args) {
+    MakeOperation<Op>(args...)->Execute();
+  }
+
+  // The gRPC objects that have to be valid until the last gRPC operation
+  // associated with this call finishes. Note that `grpc::ClientContext` is
+  // _not_ reference-counted.
+  //
+  // Important: `call_` has to be destroyed before `context_`, so declaration
+  // order matters here. Despite the unique pointer, `call_` is actually
+  // a non-owning handle, and the memory it refers to will be released once
+  // `context_` (which is owning) is released.
```

Comment:
I guess i would be more clear if you mention what `memory' exactly you mean here. Otherwise, may confused with the memory of the instance of type grpc::GenericClientAsyncReaderWriter, which is what call_ refers to but not what you mean here.

Reply:
Tried to rephrase a little bit, but this is pretty much what I mean here. The unique pointer does not actually own the `grpc::GenericClientAsyncReaderWriter` (its deletion function is a no-op). Instead, the `grpc::GenericClientAsyncReaderWriter` is allocated as part of an "arena" that gRPC releases when `grpc::ClientContext` is destroyed. I'm not sure if this is intentional or a bug.